### DISCUSSION TITLE
Option to write custom time series data in mcap files

### DISCUSSION THREAD

**Discussion by konsgken** [foxglove/mcap#1105] [General] [answered]
Hi, 

I would like to ask you if there is an option to write custom time series data, except from the ones that you provide in nuscenes2mcap repo (e.g. distance from a target vehicle through time) in MCAP files. Is there any defined message for writing custom time series data?

**Comment by jtbandes** [comment, marked as answer]
You can use any encoding for your custom data (see details of Channel and Schema records in the spec: https://mcap.dev/specification/index.html#channel-op0x04).  Common encodings are listed here: https://mcap.dev/specification/appendix.html

**Reply by konsgken** [reply]
Thanks @jtbandes  for your answer, I would like to ask one last question. I saw in the nuscenes2mcap repo that you are using JSON schemas to write time series data in MCAP files. Is there any ROS2 message  that you can use for  writing "custom" time series data (as I said in my previous question e.g. distance from a target vehicle) in MCAP files?

**Reply by jtbandes** [reply]
Maybe one of these? https://docs.ros2.org/galactic/api/std_msgs/index-msg.html

**Reply by amacneil** [reply]
To be clear, you can create your own ROS 2 message for custom time series data, and you can write _any_ message into an MCAP file.

**Reply by konsgken** [reply]
Thanks @jtbandes for the information. @amacneil is there a way to visualise (in foxglove studio) and read custom ros2 message definitions from mcap files? If yes, how can I create such a message definition;

**Reply by jtbandes** [reply]
Would recommend checking out this tutorial: https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Custom-ROS2-Interfaces.html

**Reply by amacneil** [reply]
@konsgken yes they will work in Foxglove Studio. Some panels will work with any message, other panels require a specific message.

https://foxglove.dev/docs/studio/messages/introduction
